kubespawner/proxy.py

```diff
@@ -0,0 +1,244 @@
+            lambda: safe_name in self.endpoint_reflector.endpoints,
+            'Could not find endpoints/%s after creating it' % safe_name
+        )
+
+        yield ensure_object(
+            self.core_api.create_namespaced_service,
+            self.core_api.patch_namespaced_service,
+            body=service,
+            kind='service'
+        )
+
+        yield exponential_backoff(
+            lambda: safe_name in self.service_reflector.services,
+            'Could not find service/%s after creating it' % safe_name
+        )
+
+        yield ensure_object(
+            self.extension_api.create_namespaced_ingress,
+            self.extension_api.patch_namespaced_ingress,
+            body=ingress,
+            kind='ingress'
+        )
+
+        yield exponential_backoff(
+            lambda: safe_name in self.ingress_reflector.ingresses,
+            'Could not find ingress/%s after creating it' % safe_name
+        )
+
+    @gen.coroutine
+    def delete_route(self, routespec):
+        # We just ensure that these objects are deleted.
+        # This means if some of them are already deleted, we just let it
+        # be.
+        safe_name = self.safe_name_for_routespec(routespec).lower()
+
+        delete_options = client.V1DeleteOptions(grace_period_seconds=0)
+
+        delete_endpoint = self.asynchronize(
+            self.core_api.delete_namespaced_endpoints,
+            name=safe_name,
+            namespace=self.namespace,
+        )
+
+        delete_service = self.asynchronize(
+            self.core_api.delete_namespaced_service,
+            name=safe_name,
+            namespace=self.namespace,
+        )
+
+        delete_ingress = self.asynchronize(
+                self.extension_api.delete_namespaced_ingress,
+                name=safe_name,
+                namespace=self.namespace,
+                body=delete_options,
+                grace_period_seconds=0
+        )
+
+        # This seems like cleanest way to parallelize all three of these while
+        # also making sure we only ignore the exception when it's a 404
+        def delete_if_exists(kind, future):
+            try:
+                yield future
+            except client.rest.ApiException as e:
+                if e.status != 404:
+                    raise
+                self.log.warn("Could not delete %s %s: does not exist", kind, safe_name)
+
+
+        delete_if_exists('endpoint', delete_endpoint)
+        delete_if_exists('service', delete_service)
+        delete_if_exists('ingress', delete_ingress)
+
+
+    @gen.coroutine
+    def get_all_routes(self):
+        # copy everything, because iterating over this directly is not threadsafe
+        # FIXME: is this performance intensive? It could be! Measure?
+        # FIXME: Validate that this shallow copy *is* thread safe
+        ingress_copy = dict(self.ingress_reflector.ingresses)
+        routes = {
+            i.metadata.annotations['hub.jupyter.org/proxy-routespec']:
```

Comment:
What is `i`? Each route? Original before copy? Define in a comment or docstring. Thanks 👍

Reply:
I've renamed it to 'ingress' which I think clarifies what they are (ingress objects). lmk if you think an additional doc line is still warranted!

Reply:
I've renamed it to 'ingress' now, since these are all ingress objects. Let me know if that's good enough!

Reply:
+1 to renaming. no doc line needed.